### DISCUSSION TITLE
Using r#keyword or aliases?

### DISCUSSION THREAD

**Discussion by SOF3** [rust-lang-nursery/api-guidelines#207] [General]
Rust supports a syntax called "keyword identifiers", where you may call a variable `r#type` instead of resorting to an alias like `ty`.

However, it is also common to see (even [in rust-src](https://github.com/rust-lang/rust/blob/f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec/src/libsyntax/ast.rs#L1766)) that `type` is aliased as `ty` when used in variable names or struct fields.

Should `r#type` or `ty` be prefrred? If the latter is preferred, what about other keywords? Is there a conventional mapping from the keyword to the alternative identifier?

For example, in Java, it is common to have variables named `clazz` when the variable type is a `java.lang.Class`.

It would be useful for provide such a mapping, or otherwise to encourage use of `r#` in api-guidelines (especially the naming section)

**Comment by SOF3** [comment]
I think `ty` is a special case though, because it's too common that structs are called `ty` even when they don't need to.

**Comment by jyn514** [comment]
The ones I've seen are
- `crate` -> `krate`
- `type` -> `ty`
- `self` -> `this`
- `mod` -> `module`

I don't think I've actually seen anyone using keyword identifiers unless it was done by `cargo fix` or something.

**Comment by ThomasdenH** [comment]
I like using r#keywords! Using another name is always at least slight less clear. It may be harder to r#type though, depending on the tooling.

**Comment by SOF3** [comment]
I think `r#keyword` is mostly used for backward compatibility, such as `reqwest::r#async` in the past.

**Comment by Ixrec** [comment]
Compatibility across epochs/editions is definitely the reason raw keywords / `r#keyword` were added to the language, and the use case they're designed for.

While we theoretically _could_ replace some aliases with raw keywords instead, aliases have been the only option for so long that they're the de facto standard, so we'd need a pretty strong motivation to switch. And personally, even if we could boil that ocean, I like aliases better for subjective aesthetic reasons and because they're fundamentally a simpler solution to this problem (grokking how raw keywords help with the cross-epoch lexing problem may not be that hard, but it's not trivial either).

**Comment by SOF3** [comment]
Also recall how difficult it is to figure out hwo to use raw keywords in `quote`.

**Comment by Buer-Nahida** [comment]
I like `r#keyword`, because in I using cplusplus time, I always use `k_var_name` to name the variable when is a the name for variable is a duplicate of a keyword duplicate of a keyword
